### PR TITLE
Add UNSET option to amrflags array

### PR DESCRIPTION
**Pull request by BrisaDavis**
Adding a third option to the amrflags array, UNSET. When flagging, regions are checked first. Then flag2refine and/or errest (Richardson error estimator) are called only if there are UNSET cells left in the grid patch.

### CODE REVIEW

File: src/1d/flagcheck.f [hunk 1]
```diff
@@ -20,7 +20,7 @@ subroutine flagcheck(rectflags,ilo,ihi,mbuff,iflags,
 c
 
       do 10 i = imin, imax
-        if (rectflags(i) .eq. goodpt) go to 10
+        if (rectflags(i) .le. DONTFLAG) go to 10
         if (iflags(i) .ne. 1) then  !point not nested. turn off
              rectflags(i) = 0.
```
Comment (rjleveque) [inline src/1d/flagcheck.f]: should we instead set
    `rectflags(i) = DONTFLAG`
for consistency and clarity?

File: src/2d/errf1.f [hunk 2]
```diff
@@ -63,26 +63,34 @@ subroutine errf1(rctfine,nvar,rctcrse,mptr,mi2tot,mj2tot,
...
+     .         .or. rctflg(ifine,jfine+1) == UNSET
+     .         .or. rctflg(ifine+1,jfine+1) == UNSET) then
+              xofi  = xleft + (dble(ifine) - .5d0)*hx
+              term1 = rctfine(1,ifine,jfine)
+              term2 = rctfine(1,ifine+1,jfine)
+              term3 = rctfine(1,ifine+1,jfine+1)
+              term4 = rctfine(1,ifine,jfine+1)
+c             # divide by (aval*order) for relative error
+              aval  = (term1+term2+term3+term4)/4.d0
+              est   =  dabs((aval-rctcrse(1,i,j))/ order)
```
Comment (rjleveque) [inline src/2d/errf1.f]: I'm confused about what this routine is doing (also the old version in master). At this line it seems like `rctcrse` and `rctfine` are solution values, and that is what is passed in as `alloc(locbig)` and `valbgc` from `errest`, I believe.

But in line 88 below, `rctcrse(1,i,j) = rflag` so it is being used as a flag?  

Maybe this is correct and `rctcrse` is just being temporarily used to hold the flag on the coarse grid that gets passed down to the fine grid `rctflg` in  the next loop.  If so this seems confusing. Would it be better to define a new local variable `rctflgc`, for example?  At least we need a comment or two.

Maybe @mjberger can comment.

**Comment by mjberger** [comment]
Looks ok to me.

You are right that the rctcrse array is being re-=used to hold flags using the coarse dimensions, after the arrest estimate in that array has been used. Then it is transferred to the  fine grid flagging array.   There is a lot of storage being used in this routine - fine and coarse, values and flags. In the old days there was no dynamics memory allocation in Fortran. Now there is, but not sure it's better to overuse the stack space.

 — Marsha

**Comment by BrisaDavis** [comment]
@rjleveque Additional comments added to the code for clarity. Also, small change in flagcheck.f made per your comment.

**Comment by mandli** [comment]
I am just reading through each of the PRs issued but it seems like this might change the precedence of the region flagging?  Who gets to decide if a cell is flagged as "DONTFLAG" before `flag2refine` is called in this PR?

**Comment by rjleveque** [comment]
My understanding:

Yes, regions are checked first now.

If the current level is already at least as large as the maximum `maxlevel` specified for any region that covers this cell, then I think it is forbidden to refine this cell further and so the flag is set to `DONTFLAG` and `flag2refine` doesn't have to be applied to this cell.

**Comment by BrisaDavis** [comment]
Currently on the master branch the flagging occurs as follows: 
- flglvl2 first calls flagger, which calls flag2refine and/or the errest. In AMRClaw, both of these routines flag any cells they want (not constrained by regions). In GeoClaw, flag2refine checks regions to see if a cell should be flagged, and then flags any other cells it wants (not constrained by regions).
- flglvl2 then calls bufnst2, which calls flagregions2. This has the effect of setting cells that are forbidden from being refined to goodpt (even though they might have been flagged by flag2refine).

So, in this old method flag2refine flagged whatever cells it wanted and then flagregions2 got the final say by enforcing regions flagging.

In the proposed changes the flagging would occur as follows:
- flglvl2 calls flagger, which first called flagregions2. This sets cells to DOFLAG or DONTFLAG as regions require.
- flagger then calls flag2refine and/or errest, which only flags cells that are UNSET.

So, in this new method flagregions2 enforces regions flagging first and then flag2refine (and/or errest) are only allowed to modify cells that are still UNSET.

**Comment by mjberger** [comment]
As long as bufnst can still buffer and flag cells that it needs to, it shouldn't be a problem. But have you tested it to make sure ?

Just back to reading through code,

 — Marsha

**Comment by BrisaDavis** [comment]
bufnst isn't constrained at all by UNSET. The only change to bufnst is that I moved the call to flagregions2 into flagger instead. I did make changes to projc2, which is called by bufnst -- but those changes are just to check if a flag is set to less than DONTFLAG (to allow for the fact that UNSET = -1 is less than DONTFLAG) instead of checking to see if a flag is set equal to goodpt.

All of the tests I ran worked as expected. If you have a specific test in mind, I would be happy to run it.

**Comment by mandli** [comment]
So if there are "regions" outside of the normal regions (such as the topography or storm radii criteria) those will be limited by the regions?  Does this just mean we really need to generalize the regions sooner rather than later?

**Comment by rjleveque** [comment]
@mandli: The other "regions" you mention are GeoClaw specific and relate more to https://github.com/clawpack/geoclaw/pull/318 I think?  In that code, the `flag2refine` routine still checks them, but only if `flagregions` hasn't already determined that there is no point in checking further.

**Comment by rjleveque** [comment]
After more tests and discussion with @mjberger, this seems fine and I am going to merge it and then rerun the travis tests on https://github.com/clawpack/geoclaw/pull/318.

**Pull request merged**